### PR TITLE
AD74412R/AD74413R driver

### PR DESCRIPTION
**Pull request by CiprianRegus**
Implemented a driver and demo project for the AD74413R and AD74412R ADC-DAC devices. 

Draft until better tested on STM32.

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
some comments prior to `ready for review` stage.

File: drivers/adc-dac/ad74413r/ad74413r.h [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+#define AD74413R_CH_C                   2
+#define AD74413R_CH_D                   3
+
+/** The value of the sense resistor in ohms */
+#define AD74413R_RSENSE                 100
+/** 16 bit ADC */
+#define AD74413R_ADC_MAX_VALUE		65535
+
+/** Register map */
+
```
Comment (amiclaus) [inline drivers/adc-dac/ad74413r/ad74413r.h:59]: nit: extra line

File: drivers/adc-dac/ad74413r/ad74413r.h [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+#define AD74413R_ERR_CLR_MASK			NO_OS_GENMASK(15, 0)
+#define AD74413R_SPI_CRC_ERR_MASK		NO_OS_BIT(13)
+#define AD74413R_CH_FUNC_SETUP_MASK             NO_OS_GENMASK(3, 0)
+#define AD74413R_ADC_RANGE_MASK                 NO_OS_GENMASK(7, 5)
+#define AD74413R_ADC_REJECTION_MASK             NO_OS_GENMASK(4, 3)
+#define AD74413R_DEBOUNCE_TIME_MASK             NO_OS_GENMASK(4, 0)
+#define AD74413R_DEBOUNCE_MODE_MASK             NO_OS_BIT(5)
+#define AD74413R_DIAG_RESULT_MASK		NO_OS_GENMASK(15, 0)
+#define AD74413R_REV_ID				NO_OS_GENMASK(7, 0)
+
```
Comment (amiclaus) [inline drivers/adc-dac/ad74413r/ad74413r.h:108]: nit: extra line

File: projects/ad74413r/src/platform/maxim/main.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	if (ret)
+		return ret;
+
+	maxim_uart_stdio(uart_desc);
+	ret = dummy_example_main();
+#endif
+
+	return ret;
+}
+
```
Comment (amiclaus) [inline projects/ad74413r/src/platform/maxim/main.c:73]: nit: extra line

File: drivers/adc-dac/ad74413r/ad74413r.c [hunk 4]
```diff
@@ -0,0 +1,1050 @@
...
+ * @param rejection - Rejection register value
+ * @return 0 in case of success, -EINVAL otherwise
+ */
+static int ad74413r_rejection_to_rate(enum ad74413r_rejection rejection,
+				      enum ad74413r_adc_sample *rate)
+{
+	switch (rejection) {
+	case AD74413R_REJECTION_50_60:
+		*rate = AD74413R_ADC_SAMPLE_20HZ;
+		return 0;
```
Comment (amiclaus) [inline drivers/adc-dac/ad74413r/ad74413r.c:116]: ditto.

File: drivers/adc-dac/ad74413r/ad74413r.c [hunk 5]
```diff
@@ -0,0 +1,1050 @@
...
+	if (ret)
+		return ret;
+
+	if (desc->chip_id == AD74413R)
+		delay = conv_times_ad74413r[rejection];
+	else
+		delay = conv_times_ad74412r[rejection];
+
+	/** Wait for all channels to complete the conversion. */
+	if (delay < 1000)
```
[the remark this answers is not in the document]
Comment (CiprianRegus) [reply drivers/adc-dac/ad74413r/ad74413r.c:576]: It's the amount of microseconds in a millisecond. The elements of the `conv_times_ad74413r` array are microsecond delay values, which are present in the datasheet.

**Comment by CiprianRegus** [comment]
- Implemented review change requests.
- Added the `ad74413r_reg_read_raw` function, which is used to read a raw comm frame. This will be useful for the IIO driver.

**Review by dbogdan — APPROVED** [review]

**Review by amiclaus — APPROVED** [review]

File: projects/ad74413r/builds.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "stm32": {
+    "dummy_example": {
+      "flags" : "DUMMY_EXAMPLE=y IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n"
```
Comment (rbolboac) [inline projects/ad74413r/builds.json:4]: are IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n flags needed?

File: projects/ad74413r/builds.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    }
+  },
+  "maxim": {
+    "dummy_example": {
+      "flags" : "DUMMY_EXAMPLE=y IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n TARGET=max78000"
```
Comment (rbolboac) [inline projects/ad74413r/builds.json:9]: are IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n flags needed?

File: projects/ad74413r/builds.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "dummy_example": {
+      "flags" : "DUMMY_EXAMPLE=y IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n TARGET=max32655"
```
Comment (rbolboac) [inline projects/ad74413r/builds.json:12]: are IIO_EXAMPLE=n IIO_TRIGGER_EXAMPLE=n flags needed?

File: projects/ad74413r/src.mk [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+		$(INCLUDE)/no_os_irq.h      \
+		$(INCLUDE)/no_os_list.h      \
+		$(INCLUDE)/no_os_crc8.h      \
+		$(INCLUDE)/no_os_uart.h      \
+		$(INCLUDE)/no_os_lf256fifo.h \
+		$(INCLUDE)/no_os_util.h \
+		$(INCLUDE)/no_os_units.h
+
+SRCS += $(DRIVERS)/api/no_os_gpio.c \
+		$(DRIVERS)/api/no_os_i2c.c  \
```
Comment (rbolboac) [inline projects/ad74413r/src.mk:29]: is i2c needed?

**Review by rbolboac — COMMENTED** [review]
looks good, some small questions on my side.

File: projects/ad74413r/builds.json (hunk 6 above)
Comment (CiprianRegus) [reply projects/ad74413r/builds.json]: They are not. Will add them back once the IIO example is implemented.

**Comment by CiprianRegus** [comment]
- Removed unused CFLAGS.

**Review by rbolboac — APPROVED** [review]

**Review by buha — APPROVED** [review]

**Pull request merged**